openpype/plugins/publish/integrate_new.py

```diff
@@ -110,801 +156,494 @@ class IntegrateAssetNew(pyblish.api.InstancePlugin):
-                    "files are the same {} to {}".format(src, dst)
-                )
-                os.remove(dst)
-                try:
-                    shutil.copyfile(src, dst)
-                    self.log.debug("Copying files with shutil...")
-                except OSError as e:
-                    self.log.critical("Cannot copy {} to {}".format(src, dst))
-                    self.log.critical(e)
-                    six.reraise(*sys.exc_info())
-            if str(getsize(src)) in str(getsize(dst)):
-                break
-
-    def hardlink_file(self, src, dst):
-        dirname = os.path.dirname(dst)
+        # Finalize the representations now the published files are integrated
+        # Get 'files' info for representations and its attached resources
+        anatomy = instance.context.data["anatomy"]
+        representation_writes = []
+        new_repre_names_low = set()
+        for prepared in prepared_representations:
+            transfers = prepared["transfers"]
+            representation = prepared["representation"]
+            representation["files"] = self.get_files_info(
+                transfers, sites, anatomy
+            )
 
-        try:
-            os.makedirs(dirname)
-        except OSError as e:
-            if e.errno == errno.EEXIST:
-                pass
-            else:
-                self.log.critical("An unexpected error occurred.")
-                six.reraise(*sys.exc_info())
+            # Set up representation for writing to the database. Since
+            # we *might* be overwriting an existing entry if the version
+            # already existed we'll use ReplaceOnce with `upsert=True`
+            representation_writes.append(ReplaceOne(
+                filter={"_id": representation["_id"]},
+                replacement=representation,
+                upsert=True
+            ))
+
+            new_repre_names_low.add(representation["name"].lower())
+
+        # Delete any existing representations that didn't get any new data
+        # if the instance is not set to append mode
+        if not instance.data.get("append", False):
+            delete_names = set()
+            for name, existing_repres in existing_repres_by_name.items():
+                if name not in new_repre_names_low:
+                    # We add the exact representation name because `name` is
+                    # lowercase for name matching only and not in the database
+                    delete_names.add(existing_repres["name"])
+            if delete_names:
+                representation_writes.append(DeleteMany(
+                    filter={
+                        "parent": version["_id"],
+                        "name": {"$in": list(delete_names)}
+                    }
+                ))
+
+        # Write representations to the database
+        bulk_write(representation_writes)
+
+        # Backwards compatibility
+        # todo: can we avoid the need to store this?
```

Comment:
To be able to remove it, we'd need to handle deleting representations that were tagged for deletion (hence skipped here) directly in the integrator.

Another option might be re-using the `delete` flag logic from the top of this file in the following integrators that should only operate on representations that were actually already placed in the db

Reply:
Ah - the `repre` delete flag is not just a "delete publish instance" during this publishing. But delete the actual existing representation in the database? Doesn't sound like something the Integrator should do but would love to know the particular use cases!

Reply:
no no `delete` flag tells the publishing that this particular representation (not instance!) should not be integrated and the files will be deleted in the last publishing step. It doesn't delete anything in the db. The usecase is when studio is technically rendering image sequence from their DCC, but are only integrating Prores video for editorial that was generated from them automatically. 

Heavily used in 2d animation workflows with remote people to prevent integrating of massive file sequences, that give no benefit over 4444 prores files and only take more time to transfer and more data. 

The sequence representation is used during publishing to generate prores and then marked for deletion. Hence it is skipped in integrator and a cleanup pluing will remove it from the disk.